### PR TITLE
many: improve `snap wait` command

### PR DESCRIPTION
**Pull request by mvo5**
This is a followup for #5124 that implements the correct wait handling for more go datatypes and that also adds some missing error checks and better unit tests.

### CODE REVIEW

File: cmd/snap/cmd_wait.go [hunk 1]
```diff
@@ -118,6 +165,19 @@ func (x *cmdWait) Execute(args []string) error {
 	snapName := string(x.Positional.Snap)
 	confKey := x.Positional.Key
 
+	if snapName == "godot" && confKey == "" {
```
Comment (bboozzoo) [inline cmd/snap/cmd_wait.go]: Easter egg?

File: cmd/snap/cmd_wait.go [hunk 2]
```diff
@@ -75,10 +78,46 @@ func trueish(vi interface{}) bool {
 		if v > 0 {
 			return true
 		}
+	case int8:
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply cmd/snap/cmd_wait.go]: Maybe move it to one of the `util` packages in case it comes in handy in some other code?

[the remark this answers is not in the document]
Comment (pedronis) [reply cmd/snap/cmd_wait.go]: the problem is that we have other more important places that make assumptions about what YAML or JSON unmarshaling do,  so either we change everything or doing it here in this corner is mostly unwieldy and confusing

[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: @pedronis I looked at this but I'm still hesitant, the data we look at comes via the REST api, the fact that it is internally stored as json is very disconnected from the use of the helper. 

The change is of course trivial: http://paste.ubuntu.com/p/KnfRMvsbV2/ but I find this not easier to follow than the original version - maybe a matter of opinion? Would love to hear what @niemeyer thinks. In any case, happy to go either way.

File: cmd/snap/cmd_wait.go [hunk 3]
```diff
@@ -75,10 +78,46 @@ func trueish(vi interface{}) bool {
...
+		if v > 0 {
```
Comment (zyga) [inline cmd/snap/cmd_wait.go]: Nitpick: I notice that this is not related to the PR as the conditions was there before but I would consider -1 "truish" as well and would rewrite the test to `v != 0`

File: cmd/snap/cmd_wait.go (hunk 1 above)
Comment (zyga) [reply cmd/snap/cmd_wait.go]: What's the impact of this Easter egg on the existing `godot` snap?

File: cmd/snap/cmd_wait_test.go [hunk 4]
```diff
@@ -57,3 +58,114 @@ func (s *SnapSuite) TestCmdWait(c *C) {
...
 	c.Check(n > 2, Equals, true)
 }
+
+func (s *SnapSuite) TestCmdWaitMissingConfKey(c *C) {
+	n := 0
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		n++
+	})
+
+	_, err := snap.Parser().ParseArgs([]string{"wait", "snapName"})
```
Comment (zyga) [inline cmd/snap/cmd_wait_test.go]: Can you please add tests for the Easter egg as well.

File: cmd/snap/cmd_wait.go (hunk 3 above)
Comment (pedronis) [reply cmd/snap/cmd_wait.go]: ah, I missed this

File: cmd/snap/cmd_wait.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: I simplified as suggested now.

File: cmd/snap/cmd_wait.go (hunk 3 above)
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: Indeed, sorry for that.

File: cmd/snap/cmd_wait_test.go (hunk 4 above)
Comment (mvo5) [reply cmd/snap/cmd_wait_test.go]: I removed it again.

File: cmd/snap/cmd_wait.go (hunk 1 above)
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: No impact because wait always needs "snapname" and "confkey", still probably a bit much fun, I removed it.

File: cmd/snap/cmd_wait.go [hunk 5]
```diff
@@ -65,49 +67,67 @@ func isNoOption(err error) bool {
...
+// needs to becomes a generic "trueish" helper we need to resurrect
+// the code in 306ba60edfba8d6501060c6f773235d8c994a319 (and add nil
+// to it).
+func trueishForJsonUnmarshalledValues(vi interface{}) (bool, error) {
 	switch v := vi.(type) {
+	// limited to the types that json unmarhal can produce
+	case nil:
+		return false, nil
 	case bool:
 		if v == true {
```
Comment (zyga) [inline cmd/snap/cmd_wait.go]: `return v == true, nil`

File: cmd/snap/cmd_wait.go [hunk 6]
```diff
@@ -65,49 +67,67 @@ func isNoOption(err error) bool {
...
-			return true
+			return true, nil
+		} else {
+			return false, nil
 		}
 	case json.Number:
-		if i, err := v.Int64(); err == nil && i > 0 {
-			return true
+		if i, err := v.Int64(); err == nil {
+			if i != 0 {
```
Comment (zyga) [inline cmd/snap/cmd_wait.go]: `return i != 0, nil`

File: cmd/snap/cmd_wait.go [hunk 7]
```diff
@@ -65,49 +67,67 @@ func isNoOption(err error) bool {
...
+				return true, nil
+			} else {
+				return false, nil
+			}
 		}
-		if f, err := v.Float64(); err == nil && f != 0.0 {
-			return true
+		if f, err := v.Float64(); err == nil {
+			if f != 0.0 {
```
Comment (zyga) [inline cmd/snap/cmd_wait.go]: And so on

File: overlord/devicestate/devicestate_test.go [hunk 8]
```diff
@@ -2224,3 +2224,31 @@ func (s *deviceMgrSuite) TestReloadRegistered(c *C) {
...
 }
+
+func (s *deviceMgrSuite) TestMarkSeededInConfig(c *C) {
+	st := s.state
+	st.Lock()
+	defer st.Unlock()
+
+	// not seeded -> no config is set
+	s.state.Unlock()
+	s.mgr.Ensure()
```
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go]: this will start  a first boot change as well?   that might be annoying, not sure

File: overlord/devicestate/devicestate_test.go [hunk 9]
```diff
@@ -2224,3 +2224,31 @@ func (s *deviceMgrSuite) TestReloadRegistered(c *C) {
...
+	tr := config.NewTransaction(st)
+	tr.Get("core", "seed.loaded", &seedLoaded)
+	c.Check(seedLoaded, Equals, false)
+
+	// pretend we are seeded now
+	s.state.Set("seeded", true)
+
+	// seeded -> config got updated
+	s.state.Unlock()
+	s.mgr.Ensure()
```
Comment (pedronis) [inline overlord/devicestate/devicestate_test.go]: same for registration here?

File: overlord/devicestate/devicestate_test.go (hunk 8 above)
Comment (mvo5) [reply overlord/devicestate/devicestate_test.go]: Indeed, good point. I updated the test to pretend seeding is already happening and also setting the Serial to avoid device registration.

File: cmd/snap/cmd_wait.go [hunk 10]
```diff
@@ -65,49 +67,47 @@ func isNoOption(err error) bool {
...
+// trueish takes an interface{} and returns true if the interface value
+// looks "true". For strings thats if len(string) > 0 for numbers that
+// they are != 0 and for maps/slices/arrays that they have elements.
+//
+// Note that *only* types that the json package decode with the
+// "UseNumber()" options turned on are handled here. If this ever
+// needs to becomes a generic "trueish" helper we need to resurrect
+// the code in 306ba60edfba8d6501060c6f773235d8c994a319 (and add nil
+// to it).
+func trueishForJsonUnmarshalledValues(vi interface{}) (bool, error) {
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: trueishJSON might be enough as a function name.

File: cmd/snap/cmd_wait.go [hunk 11]
```diff
@@ -65,49 +67,47 @@ func isNoOption(err error) bool {
...
-		if len(v) > 0 {
-			return true
+		return v != "", nil
+	}
+	// arrays/slices/maps
+	typ := reflect.TypeOf(vi)
+	switch typ.Kind() {
+	case reflect.Array, reflect.Slice, reflect.Map:
+		s := reflect.ValueOf(vi)
+		if s.Kind() == reflect.Ptr {
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: How can that happen? Shouldn't typ.Kind() be Ptr in that case?

File: cmd/snap/cmd_wait.go [hunk 12]
```diff
@@ -65,49 +67,47 @@ func isNoOption(err error) bool {
...
-		if len(v) > 0 {
-			return true
+		switch s.Kind() {
+		case reflect.Array, reflect.Slice, reflect.Map:
+			return s.Len() > 0, nil
 		}
 	}
-	return false
+
+	return false, fmt.Errorf("cannot test type %T for trueishness", vi)
```
Comment (niemeyer) [inline cmd/snap/cmd_wait.go]: truth? :)

File: cmd/snap/cmd_wait.go (hunk 11 above)
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: Yeah, this was from an earlier incarnation of the code, removed now.

File: cmd/snap/cmd_wait.go (hunk 12 above)
Comment (mvo5) [reply cmd/snap/cmd_wait.go]: Yes :) Thank you!
